pywinauto/win32_hooks.py

```diff
@@ -241,83 +407,143 @@ class Hook(object):
                  1012: 'Shift',
                  1013: 'Win'}
 
-    event_types = {0x100: 'key down',  # WM_KeyDown for normal keys
-                   0x101: 'key up',  # WM_KeyUp for normal keys
-                   0x104: 'key down',  # WM_SYSKEYDOWN, used for Alt key.
-                   0x105: 'key up',  # WM_SYSKEYUP, used for Alt key.
+    event_types = {win32con.WM_KEYDOWN: 'key down',     # WM_KEYDOWN for normal keys
+                   win32con.WM_KEYUP: 'key up',         # WM_KEYUP for normal keys
+                   win32con.WM_SYSKEYDOWN: 'key down',  # WM_SYSKEYDOWN, is used for Alt key.
+                   win32con.WM_SYSKEYUP: 'key up',      # WM_SYSKEYUP, is used for Alt key.
                    }
 
-    WH_KEYBOARD_LL = 0x00D
-    WH_MOUSE_LL = 0x0E
-    WM_QUIT = 0x0012
-
     def __init__(self):
-        self.handler = 0
+        self.handler = None
         self.pressed_keys = []
         self.keyboard_id = None
         self.mouse_id = None
         self.mouse_is_hook = False
         self.keyboard_is_hook = False
 
+    def _process_kbd_data(self, kb_data_ptr):
+        """Proces KBDLLHOOKSTRUCT data received from low level keyboard hook calls"""
+        kbd = KBDLLHOOKSTRUCT.from_address(kb_data_ptr)
+        current_key = None
+        key_code = kbd.vkCode
+        if key_code == win32defines.VK_PACKET:
+            scan_code = kbd.scanCode
+            current_key = six.unichr(scan_code)
+        elif key_code in self.ID_TO_KEY:
+            current_key = six.u(self.ID_TO_KEY[key_code])
+        else:
+            al = ActionLogger()
+            al.log("_process_kbd_data, bad key_code: {0}".format(key_code))
+
+        return current_key
+
+    def _process_kbd_msg_type(self, event_code, current_key):
+        """Process event codes from low level keyboard hook calls"""
+        event_type = None
+        event_code_word = 0xFFFFFFFF & event_code
+        if event_code_word in self.event_types:
+            event_type = self.event_types[event_code_word]
+        else:
+            al = ActionLogger()
+            al.log("_process_kbd_msg_type, bad event_type: {0}".format(event_type))
+
+        if event_type == 'key down':
+            self.pressed_keys.append(current_key)
+        elif event_type == 'key up':
+            if current_key in self.pressed_keys:
+                self.pressed_keys.remove(current_key)
+            else:
+                al = ActionLogger()
+                al.log("_process_kbd_msg_type, can't remove a key: {0}".format(current_key))
+
+        return event_type
+
+    def _keyboard_ll_hdl(self, code, event_code, kb_data_ptr):
+        """Execute when a keyboard low level event was catched"""
+        try:
+            # The next hook in chain must be always called
+            res = windll.user32.CallNextHookEx(self.keyboard_id,
```

Comment:
The MSDN docs say: "A hook procedure can call this function either before or after processing the hook information." Notice that the hooks run in the context of an application so, I think, the ```CallNextHookEx``` simply sends a message to another registered app (we by ourselves have the loop message) Maybe it also calls to ```LoadLibrary```.